boot: load addons from systemd-boot Type 1 entries

RaitoBezarius opened the pull request:
This PR introduces a way to load addons from systemd-boot Type 1 entries using the `addon` stanza (which can be repeated as much time as you want and is **sensitive** to **order**).

It also pick up those addons in systemd-stub and load them as part of the addon loading procedure.

<!-- devel-freezer = {"comment-id":"1627439012","freezing-tag":"v254-rc1"} -->

<!-- devel-freezer = {"comment-id":"1627445767","freezing-tag":"v254-rc2"} -->

RaitoBezarius committed:
systemd-boot: load addons from type 1 and expose them via EFI protocol

This make it possible to attach addons to any Type #1 entry
and get them back in the next bootable stage via an EFI protocol.

RaitoBezarius committed:
libefi(device-path): introduce `make_multiple_file_device_path`

This build an `EfiDevicePath**` of the form:

`FilePath(\files[0])\EndNode()FilePath(\files[1])\EndNode()…\FilePath(\files[n])\EndNode()`

which should really be read as a list:

[ FilePath(\files[0])\EndNode() ; FilePath(\files(1))\EndNode() ; … ;
FilePath(\files[n])\EndNode() ]

medhefgo commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,36 @@
+        return BS->InstallMultipleProtocolInterfaces(ret_addons_handle,
+                                            MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL),
+                                            &addons_dp);
+}
```
You're thinking too much of how the initrd loadfile stuff works.

There is no need for a (global) efi_addon_device_path struct or a new handle for it.

Just create a heap allocated mulit device path and install it onto the handle of the *loaded image*:
```
BS->LoadImage(…, loaded_image_handle);
…
EFI_DEVICE_PATH *dp = make_multiple_file_device_path(…);
…
BS->InstallMultipleProtocolInterfaces(loaded_image_handle, MAKE_GUID_PTR(SYSTEMD_ADDON_DEVICE_PATH), &dp, NULL);
```

This way, the stub can just look at the handle it receives in efi_main for the device path and there is no guesswork to be done. Also, it has to be under our own new GUID as `EFI_DEVICE_PATH_PROTOCOL_GUID` as fixed meanings defined by the UEFI spec. We cannot step on that.

RaitoBezarius replied:
Oh, I see, I was not sure, I could directly install it with my custom GUID, I can just keep the function right and pass the `loaded_image_handle` or should I just ensure I get an EFI loaded image handle to ensure no one confuses the handle?
Thank you for the pointer :)

RaitoBezarius committed:
systemd-stub(addons): load generic addons via EFI protocol

- reads from EFI protocol
- make the loading phase agnostic from the "read source"
- restructure the addon entry to be generic and not only cmdline for
  future initrd loading, etc.
- add a feature called `OFFER_ADDONS` on sd-boot to signal support for
  this EFI protocol

RaitoBezarius commented on src/boot/efi/stub.c:
```diff
@@ -393,10 +463,14 @@ static EFI_STATUS run(EFI_HANDLE image) {
+
+        if (szs[UNIFIED_SECTION_LINUX] == 0) {
+                err = EFI_NOT_FOUND;
+                log_error_status(err, "Unable to locate embedded .linux section: %m, will skip...");
```
This is a WIP for my own tests

RaitoBezarius commented on src/fundamental/efivars-fundamental.h:
```diff
@@ -20,13 +20,15 @@
 #define EFI_LOADER_FEATURE_SORT_KEY                (UINT64_C(1) << 8)
 #define EFI_LOADER_FEATURE_SAVED_ENTRY             (UINT64_C(1) << 9)
 #define EFI_LOADER_FEATURE_DEVICETREE              (UINT64_C(1) << 10)
+#define EFI_LOADER_FEATURE_ADDONS                  (UINT64_C(1) << 11)
```
Unsure about those, let me know if this is not appropriate, I will drop them

RaitoBezarius committed:
bootctl: support for `add-on` stanza in type 1

- implements cleanup
- show boot entry with addons in JSON and text

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1493,6 +1497,16 @@ static void config_entry_add_type1(
+                        entry->addons = xrealloc(
+                                entry->addons,
+                                n_addons == 0 ? 0 : (n_addons + 1) * sizeof(uint16_t *),
+                                (n_addons + 2) * sizeof(uint16_t *));
```
this expression gives me nightmares. We should have a helper for this somewhere, given this is now used multiple times.

RaitoBezarius replied:
Tried something with macros, let me know if you prefer something else in commit fa1afee615

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,70 @@
 
+EFI_STATUS make_multiple_file_device_path(EFI_HANDLE device, const char16_t **files, EFI_DEVICE_PATH
+                                          ***ret_dp)
+{
```
plesase follow our usual coding style, i.e. line break after the `(`, then indent to 16spaces, then put `{` on the same line as `(`

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,70 @@
+        assert(ret_dp);
+
+        err = BS->HandleProtocol(device, MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL), (void **)
+                                 &original_device_path);
```
we line break at 109ch. see .editorconfig and related files

RaitoBezarius replied:
Hmm, my editor linebreaked at 109ch according to editorconfig, but do you prefer to line break the whole expr `(void **) &...` ?

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,70 @@
+        while (!device_path_is_end(end_node))
+                end_node = device_path_next_node(end_node);
+
+
```
spurious double empty line

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,70 @@
+        size_t o_dp_size = (uint8_t *) end_node - (uint8_t *) original_device_path;
+        STRV_FOREACH(file, files) {
+                n_files += 1;
+        }
```
no {} around single line blocks

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,70 @@
+                size_t file_size = strsize16(*file);
+
+                // 1st element: FILEPATH_DEVICE_PATH + path name payload
+                // 2nd element: DEVICE_PATH_END_NODE
```
as per coding style, use `/* */` style comments for explanatory descriptions, and `//` for FIXME style comments. this one is the former

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -136,3 +200,95 @@ EFI_DEVICE_PATH *device_path_replace_node(
+        EFI_DEVICE_PATH *ret_dp = NULL;
+        size_t n_dps = 0;
+        size_t n_instances = 0;
+        size_t total_dp_size = 0;
```
merge these three lines

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -136,3 +200,95 @@ EFI_DEVICE_PATH *device_path_replace_node(
+        const EFI_DEVICE_PATH *end = dp;
+        const EFI_DEVICE_PATH **instances = NULL;
+        EFI_DEVICE_PATH *next = NULL;
+        EFI_DEVICE_PATH *ret_dp = NULL;
```
merge these two lines

poettering replied:
also, please use `ret_` as prefix exclusively for return parameters of function, nothing else, as per coding style

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -136,3 +200,95 @@ EFI_DEVICE_PATH *device_path_replace_node(
+         * (5) return an array of EFI_DEVICE_PATH*, i.e. EFI_DEVICE_PATH**
+         * */
+        const EFI_DEVICE_PATH *end = dp;
+        const EFI_DEVICE_PATH **instances = NULL;
```
merge these two lines

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -136,3 +200,95 @@ EFI_DEVICE_PATH *device_path_replace_node(
+
+        /* We perform the copy mentioned in (2) */
+        ret_dp = xmalloc(total_dp_size);
+        mempcpy(ret_dp, dp, total_dp_size);
```
if you don't intend to use the return param, use simple memcpy() rather than mempcpy().

we really should add xmemdup(), similar to the memdup() we have in userspace, but backed by xmalloc()

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -136,3 +200,95 @@ EFI_DEVICE_PATH *device_path_replace_node(
+                assert(instance_index < n_instances);
+
+                instances[instance_index] = end;
+                log_internal(EFI_SUCCESS, "wrote %lu-instance, now at %p", instance_index, end);
```
output size_t via `%zu`

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -136,3 +200,95 @@ EFI_DEVICE_PATH *device_path_replace_node(
+
+                if (device_path_is_end(end))
+                        break;
+                else
```
the `else` is unnecessary, please remove

poettering commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,26 @@
+#include "log.h"
+
+EFI_STATUS addons_install(EFI_LOADED_IMAGE_PROTOCOL *stub_image, const char16_t **addons)
+{
```
coding style, please `{` on the end of previous line

poettering commented on src/boot/efi/addon-util.h:
```diff
@@ -0,0 +1,23 @@
+        return (dp->Header.Type == MEDIA_DEVICE_PATH &&
+                dp->Header.SubType == MEDIA_VENDOR_DP &&
+                memcmp(&dp->Guid, MAKE_GUID_PTR(SYSTEMD_ADDON_MEDIA), sizeof(EFI_GUID)) == 0
+               );
```
the `()` are unnecessary

poettering commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,26 @@
+}
+
+EFI_STATUS walk_addons_in_device_path(EFI_DEVICE_PATH *addons, char16_t **files)
+{
```
coding style

poettering commented on src/boot/efi/stub.c:
```diff
@@ -21,6 +22,14 @@
 _used_ _section_(".sdmagic") static const char magic[] = "#### LoaderInfo: systemd-stub " GIT_VERSION " ####";
 
+struct addon_entry {
+        char16_t        *source_path;
```
weird spaces?

poettering commented on src/boot/efi/stub.c:
```diff
@@ -21,6 +22,14 @@
+        const EFI_DEVICE_PATH *device_path;
+};
+static int cmp_addon_entry(const struct addon_entry *s1, const struct addon_entry *s2) {
+        return strcmp16((const char16_t*)&s1->source_path, (const char16_t*)&s2->source_path);
```
please compare by basename, not by path

RaitoBezarius replied:
@medhefgo does all EFI_DEVICE_PATH have a meaning for basename based on their string representation?

RaitoBezarius replied:
I'd see this problem as the "other face" of the problem mentioned regarding multiple initrds ordering anyway: https://github.com/systemd/systemd/pull/28070#issuecomment-1596761524

Maybe, all of this can be sidestepped by comparing for `.fname` section if it exists?

poettering replied:
well, i'd suppress all add-ons which carry a .fname where the actual filename doesn't match the .fname data. for all add-ons that pass this (i.e. either have a matching filename or don't have a PE section .fname) we should order by filename.

medhefgo replied:
Argh. Stop adding new sections for ever little thing. *every* section gets padded to 512 bytes on disk and to a full page when loaded into memory. This is stupidly wasteful fore a simple file name.

Just declare a new structured data section for this kind of stuff.

bluca replied:
as mentioned in TODO we should expand https://systemd.io/ELF_PACKAGE_METADATA/ to cover PE too, and that could be a new field in there

poettering replied:
@bluca for add-ons the ELF package metadata doens't really make sense does it, because typically locally generated, not packaged.

RaitoBezarius replied:
Biggest unresolved point is this one.

RaitoBezarius replied:
I introduced c0250f4d5a5aaf57d7e13303bee073545f17c600 to attempt to tackle this, can you let me know if this is the right direction or not?

RaitoBezarius replied:
I wonder if it should be a JSON structure so that it can be extended by external consumers.

RaitoBezarius replied:
(but JSON parser in UEFI… meh… .o O (ASN.1? :D))

bluca replied:
> @bluca for add-ons the ELF package metadata doens't really make sense does it, because typically locally generated, not packaged.

I imagine we'll start using it for more things, like for UKIs. You are always going on and on and on about how we don't validate filenames for anything after all, so it will likely apply to absolutely everything :>

bluca replied:
but maybe @medhefgo has a point, if it's not intended to be for "public" consumption, but just for internal use and sanity check, perhaps we could just use some data structure instead of adding a new header section?

poettering commented on src/boot/efi/stub.c:
```diff
@@ -88,6 +97,7 @@ static void export_variables(EFI_LOADED_IMAGE_PROTOCOL *loaded_image) {
                 EFI_STUB_FEATURE_PICK_UP_SYSEXTS |          /* We pick up system extensions from the boot partition */
                 EFI_STUB_FEATURE_THREE_PCRS |               /* We can measure kernel image, parameters and sysext */
                 EFI_STUB_FEATURE_RANDOM_SEED |              /* We pass a random seed to the kernel */
+                EFI_STUB_FEATURE_PICK_UP_ADDONS |           /* We load addons provided via EFI */
```
i think we should advertise which sections precisely we pick up via addons, i.e. distinct flags for this. as we add support for more userspace wants to know which ones

RaitoBezarius replied:
so `EFI_STUB_FEATURE_PICK_UP_INITRD`, CMDLINE, etc, right?

RaitoBezarius replied:
Split into 2 pick up features, 1 feature for sd-boot called "offer addons"

poettering commented on src/boot/efi/stub.c:
```diff
@@ -228,33 +241,88 @@ static EFI_STATUS load_addons_from_dir(
 
+static EFI_STATUS load_addons_from_efi(EFI_LOADED_IMAGE_PROTOCOL *image, struct addon_entry **addons, size_t *n_items, size_t
+                                       *n_allocated)
+{
```
coding style...

poettering commented on src/boot/efi/stub.c:
```diff
@@ -393,10 +462,14 @@ static EFI_STATUS run(EFI_HANDLE image) {
+        if (err != EFI_SUCCESS) {
+                return log_error_status(err, "Unable to locate embedded sections in the current binary: %m");
+        }
+
```
coding style, no {} around single-line if blocks

poettering reviewed (changes requested):
some superficial comments

RaitoBezarius committed:
libefi(util): introduce `xrealloc_extra_(k)_item(s)`

This is helpful to extend of K items an existing NULL-terminated array,
by default, we use it with an extension of K=1, in `boot.c` for loading a list of initrds
and addons.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -576,6 +578,8 @@ static void print_status(Config *config, char16_t *loaded_image_path) {
                 STRV_FOREACH(initrd, entry->initrd)
                         printf("        initrd: %ls\n", *initrd);
+                STRV_FOREACH(addon, entry->addons)
+                        printf("         addon: %ls\n", *addon);
```
addon or add-on?

in the output here you spell it `addon` but in the type1 files you parse `add-on`. stick to one spelling. so far @bluca used `addon` mostly. I actually prefer `add-on`, but i guess it doesn't matter, hence maybe use `addon` everywhere then

RaitoBezarius replied:
I prefer `add-on` too, but I don't want to spend bikeshedding time on that, feel free to decide for me :)

medhefgo replied:
The correct spelling is "Add-on" when in object form and "add on" when in verb form. Just like "(a) Log-in" vs "(to) log in".

bluca replied:
I prefer 'addon' so that it can always be used, if it was 'add-on' then every time it's used in code (like here) it would need to become 'add_on' give '-' is not a legal variable character

poettering replied:
that thinking sounds ok to me. so let's just declare that henceforth the canonical spelling in code and cnfig files of this concept shall be "addon",

RaitoBezarius replied:
Done

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,67 @@
+
+        size_t o_dp_size = (uint8_t *) end_node - (uint8_t *) original_device_path;
+        STRV_FOREACH(file, files)
+                n_files += 1;
```
hmm, we should add a strv_length() helper that does this to src/fundamental/, similar to the call we have in userspace

RaitoBezarius replied:
Done

poettering commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,67 @@
+        STRV_FOREACH(file, files)
+                n_files += 1;
+
+        *ret_dp = xnew(EFI_DEVICE_PATH*, n_files);
```
across our codebase we generally try to ensure that it is OK to encode an empty object as NULL.

Hence, maybe do here:

```c
if (n_files == 0) {
        *ret_dep = NULL;
        return EFI_SUCCESS;
}
```

*ret_dp = xnew(…);
```

RaitoBezarius replied:
Done

poettering commented on src/boot/efi/util.h:
```diff
@@ -52,6 +52,10 @@ static inline void *xrealloc(void *p, size_t old_size, size_t new_size) {
 
+#define xrealloc_extra_k_items(type, p, current_size, k) xrealloc(p, \
+                current_size == 0 ? 0 : (current_size + 1) * sizeof((type *)), \
+                (current_size + k + 1) * sizeof((type *)))
```
i don't like the amount of double-eval/side-effects. Can't we do better, maybe with a static inline helper?

also misses a couple of () to avoid unexpected operator resolution order

RaitoBezarius replied:
Here's a suggestion with static inline helper, we lose the `type*` on the extra_k_items, but seems fine.

RaitoBezarius committed:
libefi(util): introduce `strv_length` for char16_t

With improved const-correctness across the previous users.

RaitoBezarius committed:
boot(pe): introduce `.binrel` as PE metadata

Following
https://github.com/systemd/systemd/pull/28057/files#r1233760527,

we realized the need for EFI metadata, as EFI device paths will not
always carry the filename information.

Here is an attempt to materialize it as a PE section `.sdmeta` to avoid
piling up many sections which will be all aligned to full pages in
memory and 512 bytes on disk.

This is a structured data section for any PE metadata, it starts with a
`filename` field.

In a future PR, stubs, addons and systemd-boot will make use of that
section.

bluca commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,67 @@
 #include "util.h"
 
+EFI_STATUS make_multiple_file_device_path(
+                EFI_HANDLE device, const char16_t **files, EFI_DEVICE_PATH ***ret_dp) {
```
if you line break then one param per line, and empty line after the bracket

RaitoBezarius replied:
Done.

bluca commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,67 @@
+        while (!device_path_is_end(end_node))
+                end_node = device_path_next_node(end_node);
+
+        size_t o_dp_size = (uint8_t *) end_node - (uint8_t *) original_device_path;
```
does this need some overflow check?

RaitoBezarius replied:
that the resulting size fits inside `size_t` or the operation makes somewhat sense? I can be more defensive in any case, I guess

bluca commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,42 @@
+#include "addon-util.h"
```
SPDX line missing

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,42 @@
+EFI_STATUS addons_install_proto(EFI_LOADED_IMAGE_PROTOCOL *stub_image, char16_t * const *addons) {
+        EFI_STATUS err;
+        EFI_DEVICE_PATH **dps;
+
```
assert(stub_image)

RaitoBezarius replied:
Done.

bluca commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,67 @@
+
+        size_t o_dp_size = (uint8_t *) end_node - (uint8_t *) original_device_path;
+
+        if (n_files == 0) {
```
this should probably be the first check, above

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,42 @@
+        err = make_multiple_file_device_path(stub_image->DeviceHandle, addons, &dps);
+        if (err != EFI_SUCCESS)
+                return err;
+
```
return early if dps is null

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/addon-util.c:
```diff
@@ -0,0 +1,42 @@
+        EFI_STATUS err;
+        EFI_DEVICE_PATH *dps;
+
+        if (!*addons)
```
either check or assert on addons

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/boot.c:
```diff
@@ -3,6 +3,7 @@
 #include "bcd.h"
 #include "bootspec-fundamental.h"
 #include "console.h"
+#include "addon-util.h"
```
sort alphabetically

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/boot.c:
```diff
@@ -2327,6 +2342,11 @@ static EFI_STATUS initrd_prepare(
 }
 
+static void cleanup_loaded_image(EFI_LOADED_IMAGE_PROTOCOL **loaded_image) {
+        (void) addons_unload_proto((EFI_HANDLE *)*loaded_image);
```
assert or check loaded_image

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/meson.build:
```diff
@@ -232,6 +232,7 @@ endif
 libefi_sources = files(
         'console.c',
         'device-path-util.c',
+        'addon-util.c',
```
sort alphabetically

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/stub.c:
```diff
@@ -2,6 +2,7 @@
 
 #include "cpio.h"
 #include "device-path-util.h"
+#include "addon-util.h"
```
sort

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/stub.c:
```diff
@@ -21,6 +22,14 @@
+        char16_t *source_path;
+        const EFI_DEVICE_PATH *device_path;
+};
+static int cmp_addon_entry(const struct addon_entry *s1, const struct addon_entry *s2) {
```
line break between the two

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/stub.c:
```diff
@@ -21,6 +22,14 @@
+        const EFI_DEVICE_PATH *device_path;
+};
+static int cmp_addon_entry(const struct addon_entry *s1, const struct addon_entry *s2) {
+        return strcmp16((const char16_t*)&s1->source_path, (const char16_t*)&s2->source_path);
```
assert or check s1/s2

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/stub.c:
```diff
@@ -230,33 +242,89 @@ static EFI_STATUS load_addons_from_dir(
+                struct addon_entry **addons,
+                size_t *n_items,
+                size_t *n_allocated) {
+        EFI_STATUS err;
```
line break

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/stub.c:
```diff
@@ -230,33 +242,89 @@ static EFI_STATUS load_addons_from_dir(
+                size_t *n_allocated) {
+        EFI_STATUS err;
+        EFI_DEVICE_PATH **addon_paths = NULL;
+        err = BS->HandleProtocol(image->DeviceHandle, MAKE_GUID_PTR(SYSTEMD_ADDON_MEDIA), (void **) &addon_paths);
```
line break

RaitoBezarius replied:
Done

bluca commented on src/boot/efi/stub.c:
```diff
@@ -230,33 +242,89 @@ static EFI_STATUS load_addons_from_dir(
                 char16_t **cmdline_append) {
-
-        _cleanup_(strv_freep) char16_t **items = NULL;
+        // TODO: figure out how to free and how to allocate?
```
CLEANUP_ARRAY probably

RaitoBezarius replied:
tried something, lmk

bluca commented on src/boot/efi/stub.c:
```diff
@@ -469,6 +539,7 @@ static EFI_STATUS run(EFI_HANDLE image) {
         err = cmdline_append_and_measure_addons(
                         image,
                         loaded_image,
+                        true,
```
This is a bit strange - loading from the protocol is grafted onto one of the filesystem load calls. Instead, things should be refactored so that they are 3 individual calls, so that they are nicely separated and individual.

RaitoBezarius replied:
With your new PR, I moved the stuff inside the `load_addons` call and it calls `load_addons_from_efi`, let me know if you still think I should split it in the main entrypoint and make it more visible.

bluca commented on src/boot/efi/util.h:
```diff
@@ -52,6 +52,16 @@ static inline void *xrealloc(void *p, size_t old_size, size_t new_size) {
 
+static inline void **xrealloc_extra_k_items(void *p, size_t item_size, size_t current_size, size_t extra_elements) {
+        size_t new_size = (current_size + extra_elements + 1) * item_size;
+        if (current_size == 0) {
```
no brackets around single line ifs

RaitoBezarius replied:
Done.

bluca commented on src/fundamental/string-util-fundamental.h:
```diff
@@ -102,6 +102,18 @@ static inline void *memory_startswith(const void *p, size_t sz, const sd_char *t
+#if SD_BOOT
+static inline size_t strv_length(char16_t * const *l) {
+        size_t n = 0;
+
```
return 0 early if l is null

RaitoBezarius replied:
Done.

val4oss commented on src/boot/efi/stub.c:
```diff
@@ -237,18 +256,76 @@ static EFI_STATUS load_addons_from_dir(
+                }
+
+                addon_spath = xasprintf("%ls\\%ls", prefix, d);
+                err = make_file_device_path(current_image, addon_spath, &addon_path);
```
Should be `DeviceHandle` instead of `current_image`
```suggestion
                err = make_file_device_path(current_image->DeviceHandle, addon_spath, &addon_path);
```

val4oss commented on src/boot/efi/stub.c:
```diff
@@ -237,18 +256,76 @@ static EFI_STATUS load_addons_from_dir(
+
+        err = BS->HandleProtocol(image->DeviceHandle, MAKE_GUID_PTR(SYSTEMD_ADDON_MEDIA), (void **) &addon_paths);
+
+        if (err == EFI_NOT_FOUND)
```
From the EFI protocole documentation, `HandleProtocole` doesn't return EFI_NOT_FOUND. It can only return one of theses: EFI_SUCCESS, EFI_UNSUPPORTED or EFI_INVALID_PARAMETER.
In this case it should be `EFI_UNSUPPORTED`
```suggestion
        if (err == EFI_UNSUPPORTED)
```

val4oss commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,71 @@
+        size_t o_dp_size = (uint8_t *) end_node - (uint8_t *) original_device_path;
+
+
+        *ret_dp = xnew(EFI_DEVICE_PATH*, n_files);
```
```suggestion
        *ret_dp = xnew(EFI_DEVICE_PATH*, n_files + 1);
```

val4oss commented on src/boot/efi/device-path-util.c:
```diff
@@ -3,6 +3,71 @@
+                *cur_dp = DEVICE_PATH_END_NODE;
+
+                iterator_dp++;
+        }
```
It may be good to finish the list of EFI_DEVICE_PATH by a NULL pointer, otherwise, the `while` loop in `load_addons_from_efi` is wrong.

 ```suggestion
         }
        *iterator_dp = NULL;
```

val4oss commented on src/boot/efi/util.h:
```diff
@@ -57,6 +57,15 @@ static inline void* xmemdup(const void *p, size_t l) {
+                return (void**)xrealloc(p, (current_size + 1) * item_size, new_size);
+}
+
+#define xrealloc_extra_item(type, p, current_size) ((type **) xrealloc_extra_k_items(p, current_size, sizeof(type), 1))
```
Wrong definition argument
```suggestion
#define xrealloc_extra_item(type, p, current_size) ((type **) xrealloc_extra_k_items(p, sizeof(type *), current_size, 1))
```

val4oss commented on src/boot/efi/stub.c:
```diff
@@ -388,7 +467,11 @@ static EFI_STATUS load_addons(
+        if (err != EFI_SUCCESS)
+                return err;
+
+        err = load_addons_from_efi(loaded_image, &addons, &n_items, &n_allocated);
```
This function will be called twice, due to two calls of `load_addons` from the function `run`. One call for Global addons and a second for uki addons.
Solution would be to add bool in argument, and conditionally call `load_addons_from_efi` if argument to true.

val4oss replied:
An idea to fix that: 

```diff
diff --git a/src/boot/efi/stub.c b/src/boot/efi/stub.c
index 1e7ead8b55..69b913ddb2 100644
--- a/src/boot/efi/stub.c
+++ b/src/boot/efi/stub.c
@@ -437,6 +437,7 @@ static void items_free(char16_t **items, size_t n_items) {
 static EFI_STATUS load_addons(
                 EFI_HANDLE stub_image,
                 EFI_LOADED_IMAGE_PROTOCOL *loaded_image,
+                bool load_from_efi,
                 const char16_t *prefix,
                 const char *uname,
                 char16_t **ret_cmdline,
@@ -481,9 +482,11 @@ static EFI_STATUS load_addons(
         if (err != EFI_SUCCESS)
                 return err;
 
-        err = load_addons_from_efi(loaded_image, &addons, &n_items, &n_allocated);
-        if (err != EFI_SUCCESS)
-                return err;
+        if (load_from_efi) {
+                err = load_addons_from_efi(loaded_image, &addons, &n_items, &n_allocated);
+                if (err != EFI_SUCCESS)
+                        return err;
+        }
 
         if (n_items == 0)
                 return EFI_SUCCESS; /* Empty directory */
@@ -634,6 +637,7 @@ static EFI_STATUS run(EFI_HANDLE image) {
         err = load_addons(
                         image,
                         loaded_image,
+                        true,
                         u"\\loader\\addons",
                         uname,
                         &cmdline_addons_global,
@@ -650,6 +654,7 @@ static EFI_STATUS run(EFI_HANDLE image) {
                 err = load_addons(
                                 image,
                                 loaded_image,
+                                false,
                                 dropin_dir,
                                 uname,
                                 &cmdline_addons_uki,
```

val4oss reviewed (changes requested):
Some minors issues fixed

val4oss commented on src/boot/efi/stub.c:
```diff
@@ -237,18 +256,76 @@ static EFI_STATUS load_addons_from_dir(
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to load addons from EFI protocol: %m");
+
+        while (*addon_paths) {
```
I think, We could be stricter on the condition.

```suggestion
        while (*addon_paths && (*addons_path)->Length > 0 && (*addon_paths)->tYPE != 0) {
```
Adapting the add-ons feature in the grub2 bootloader, I was stuck in this loop quite a few times
